File: musickit/src/main/java/com/druware/musickit/MusicKitHost.kt
package com.druware.musickit

import android.media.AudioManager
import android.net.Uri
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.concurrent.TimeoutException
import kotlin.time.Duration.Companion.seconds

/**
 * Apple Music for an Android application: MusicKit JS driven inside a WebView nobody ever sees.
 *
 * **What this is.** The surface a native MusicKit app uses — authorization, library playlists and
 * their tracks, catalog search, playlist creation, and a player — over MusicKit for the Web.
 * Everything Apple insists happens in a browser (sign-in, DRM, playback) happens in the hidden page;
 * everything an Android application wants to hold (models, state, transport) is here.
 *
 * **Lifetime.** The page is a 1x1 child of the activity's content view, so a host lives no longer
 * than the activity it was created against. Call [close] from the activity's own teardown.
 *
 * **Threading.** Every member may be called from any coroutine; WebView work is marshalled to the
 * main thread internally.
 *
 * **Credentials.** The developer token is held in memory and sent to the page and nowhere else. The
 * Music User Token never crosses the bridge at all: `authorize()` resolves to it inside the page,
 * and only a boolean comes back. Neither is ever put in [diagnostic].
 */
class MusicKitHost private constructor(
    private val options: MusicKitOptions,
    private val webView: HostedWebView,
    private val bridge: MusicKitBridge,
    private val diagnostics: MutableSharedFlow<String>,
) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val policy = TokenRotationPolicy(options.rotationLeadTime)

    /** Serialises rotations, so two callers arriving together do not each replace the token. */
    private val rotationGate = Mutex()

    private val authorization = MutableStateFlow(MusicAuthorizationStatus.NotDetermined)

    @Volatile
    private var closed = false

    /**
     * Human-readable diagnostic lines.
     *
     * Never carries a token, and never a URI beyond its scheme and host. The most recent
     * [MAX_BUFFERED_LINES] lines are replayed to each new collector, because nobody can collect
     * before [create] returns and startup is where the interesting failures are.
     */
    val diagnostic: SharedFlow<String> = diagnostics.asSharedFlow()

    /** Whether the user has let this application use their Apple Music account. */
    val authorizationStatus: StateFlow<MusicAuthorizationStatus> = authorization.asStateFlow()

    /** The storefront catalog requests are made against, e.g. `us`. */
    @Volatile
    var storefrontId: String? = null
        private set

    /** When the developer token in use expires, or null before the first one is fetched. */
    @Volatile
    var developerTokenExpiresAt: Instant? = null
        private set

    /** The transport. */
    val player: MusicKitPlayer = MusicKitPlayer(
        this,
        bridge,
        options.activity.getSystemService(AudioManager::class.java),
        ::diagnose,
    )

    /**
     * Opens Apple's sign-in popup and waits for the user to finish with it.
     *
     * @return the authorization status the user left behind.
     */
    suspend fun requestAuthorization(): MusicAuthorizationStatus {
        try {
            setAuthorization(bridge.authorize(), initial = false)
        } finally {
            // Whichever way the call went, the popup's work is done. onCloseWindow is the popup's
            // own signal and is not guaranteed to fire, so this is the one that always does.
            webView.dismissAuthPopup()
        }

        return authorization.value
    }

    /** Drops the Music User Token, signing the user out. */
    suspend fun unauthorize() {
        setAuthorization(bridge.unauthorize(), initial = false)

        // Belt and braces. MusicKit has just dropped its own token; clearing the origin's storage
        // as well is what guarantees nothing of this sign-in survives to be picked up when a
        // different Apple ID signs in on the same device.
        webView.clearOriginStorage()
    }

    /**
     * Lists every playlist in the user's library.
     *
     * @return the playlists, in Apple's order.
     * @throws MusicKitException Apple refused the request.
     */
    suspend fun getLibraryPlaylists(): List<MusicPlaylist> {
        val playlists = mutableListOf<MusicPlaylist>()
        forEachPage("/v1/me/library/playlists", "load your library playlists") { page ->
            playlists.addAll(AppleJson.readPlaylists(page))
        }

        return playlists
    }

    /**
     * Fetches one library playlist by its identifier.
     *
     * @param id the library playlist identifier, which starts with `p.`.
     * @return the playlist, or null when the library does not have it.
     * @throws MusicKitException Apple refused the request for a reason other than "not found".
     */
    suspend fun getLibraryPlaylist(id: String): MusicPlaylist? {
        require(id.isNotBlank()) { "A playlist identifier is required." }

        val answer = bridge.apiGet("/v1/me/library/playlists/${Uri.encode(id)}", null)
        if (answer.status == 404) {
            return null
        }

        return AppleJson.readPlaylists(requireBody(answer, "find that playlist")).firstOrNull()
    }

    /**
     * Lists the tracks of a library playlist.
     *
     * @param playlistId the library playlist identifier.
     * @return the tracks, in playlist order.
     * @throws MusicKitException Apple refused the request.
     */
    suspend fun getLibraryPlaylistTracks(playlistId: String): List<MusicSong> {
        require(playlistId.isNotBlank()) { "A playlist identifier is required." }

        val songs = mutableListOf<MusicSong>()
        forEachPage(
            "/v1/me/library/playlists/${Uri.encode(playlistId)}/tracks",
            "load that playlist's tracks",
        ) { page ->
            songs.addAll(AppleJson.readSongs(page))
        }

        return songs
    }

    /**
     * Searches the Apple Music catalog for songs.
     *
     * Needs the developer token only, so it works before the user has signed in.
     *
     * @param term what to search for.
     * @param limit how many songs to ask for; Apple's own maximum is 25.
     * @return the songs found, in Apple's order.
     * @throws MusicKitException Apple refused the request.
     */
    suspend fun searchCatalogSongs(term: String, limit: Int): List<MusicSong> {
        require(term.isNotBlank()) { "A search term is required." }
        require(limit > 0) { "A limit of at least one song is required." }

        val answer = bridge.apiGet(
            "/v1/catalog/${storefrontId ?: "us"}/search",
            mapOf("term" to term, "types" to "songs", "limit" to limit.toString()),
        )

        return AppleJson.readSearchSongs(requireBody(answer, "search the catalogue"))
    }

    /**
     * Creates a playlist in the user's library from catalog songs.
     *
     * `music.api.music` offers no POST of its own, so the page makes this one with a plain `fetch`
     * from inside itself — where the Music User Token already is, and so where it stays.
     *
     * @param name the name to give the playlist.
     * @param catalogSongIds the catalog identifiers to put in it, in order.
     * @return the new playlist's library identifier.
     * @throws MusicKitException Apple refused the request, or answered without an identifier.
     */
    suspend fun createLibraryPlaylist(name: String, catalogSongIds: List<String>): String {
        require(name.isNotBlank()) { "A playlist name is required." }

        val request = buildJsonObject {
            put("attributes", buildJsonObject { put("name", name) })
            put(
                "relationships",
                buildJsonObject {
                    put(
                        "tracks",
                        buildJsonObject {
                            put(
                                "data",
                                buildJsonArray {
                                    catalogSongIds.forEach { id ->
                                        add(
                                            buildJsonObject {
                                                put("id", id)
                                                put("type", "songs")
                                            },
                                        )
                                    }
                                },
                            )
                        },
                    )
                },
            )
        }

        val answer = bridge.apiPost("/v1/me/library/playlists", request.toString())
        diagnose("create playlist: HTTP ${answer.status ?: "?"} via ${answer.via ?: "?"}")

        val created = AppleJson.readPlaylists(requireBody(answer, "create that playlist")).firstOrNull()
        return created?.id
            ?: throw MusicKitException(
                "The playlist was created but Apple Music did not say what its identifier is.",
                "NO_PLAYLIST_ID",
            )
    }

    /**
     * Asks the page whether this device's WebView can start Widevine EME at all.
     *
     * Not in the desktop library, and here because Android needs it: some OEM WebView builds carry
     * no Widevine CDM, and without this the only symptom is that playback never starts. A host can
     * use it to say so plainly rather than appearing to hang.
     *
     * @return true when the page could obtain Widevine key-system access.
     */
    suspend fun probeWidevine(): Boolean {
        val probe = bridge.probeWidevine()
        diagnose(
            if (probe.supported) {
                "Widevine is available (${probe.keySystem ?: "unnamed key system"})"
            } else {
                "Widevine is not available: ${probe.error ?: "no reason given"}"
            },
        )

        return probe.supported
    }

    /** Closes the page and the WebView behind it. Idempotent. */
    suspend fun close() {
        if (closed) {
            return
        }

        closed = true
        scope.cancel()

        bridge.onPageEvent = null
        bridge.onLog = null
        bridge.failAll("The MusicKit host was disposed.")

        webView.onMessage = null
        webView.onRendererGone = null
        webView.onAuthPopupFailed = null
        webView.close()
    }

    /**
     * Runs one expression in the page and returns its result as JSON.
     *
     * Internal, and there for one reason: MusicKit JS is Apple's moving target, so the instrumented
     * test has to be able to ask the live page what it actually does rather than assume. A host has
     * no business here; every supported operation has a typed method above.
     *
     * @param script the expression to evaluate.
     */
    internal suspend fun evaluate(script: String): String = webView.evaluate(script)

    /**
     * Replaces the developer token, whether or not one is due.
     *
     * Internal because a host has no reason to ask for this: rotation is decided by
     * [TokenRotationPolicy]. Exposed at all so a test can exercise the path without waiting an hour.
     */
    internal suspend fun rotateNow() {
        rotationGate.withLock { rotateCore() }
    }

    /**
     * Replaces the developer token when the policy says the moment has come.
     *
     * @param trigger what is about to happen.
     * @return true when a rotation happened, so the caller knows MusicKit's queue is now empty.
     */
    internal suspend fun rotateIfNeeded(trigger: RotationTrigger): Boolean {
        if (!policy.shouldRotate(trigger, player.state.value, developerTokenExpiresAt, Instant.now())) {
            return false
        }

        return rotationGate.withLock {
            // Re-checked under the gate: another caller may have rotated while this one waited.
            if (!policy.shouldRotate(trigger, player.state.value, developerTokenExpiresAt, Instant.now())) {
                return@withLock false
            }

            rotateCore()
            true
        }
    }

    /**
     * Puts one line on [diagnostic].
     *
     * @param line the line to report. Must never carry a token or a URI beyond scheme and host.
     */
    internal fun diagnose(line: String) {
        diagnostics.tryEmit(line)
    }

    private suspend fun rotateCore() {
        val token = options.developerTokenProvider.getToken(forceRefresh = true)
        val configured = bridge.reconfigure(token.token)

        developerTokenExpiresAt = token.expiresAt
        storefrontId = configured.storefrontId ?: storefrontId
        setAuthorization(configured.isAuthorized, initial = false)

        diagnose(
            "developer token rotated; expires ${utc(token.expiresAt)}, " +
                "authorized ${configured.isAuthorized}, " +
                "instance took the new token: ${configured.developerTokenMatches}",
        )
    }

    /** Walks a paged Apple collection, handing over one response body per page. */
    private suspend fun forEachPage(path: String, what: String, onPage: (JsonElement) -> Unit) {
        var nextPath = path
        var parameters = mapOf("limit" to PAGE_SIZE.toString())

        repeat(MAX_PAGES) {
            val body = requireBody(bridge.apiGet(nextPath, parameters), what)
            onPage(body)

            val next = AppleJson.readNext(body)
            if (next.isNullOrEmpty()) {
                return
            }

            val split = AppleJson.splitNext(next)
            nextPath = split.path
            parameters = split.parameters
        }

        throw MusicKitException(
            "Apple Music kept asking for another page while trying to $what, so the request was " +
                "abandoned.",
            "TOO_MANY_PAGES",
        )
    }

    private fun setAuthorization(authorized: Boolean, initial: Boolean) {
        // Before anyone has been asked, "not authorized" and "not determined" are the same
        // observation; only an explicit authorize or unauthorize can tell them apart.
        authorization.value = when {
            authorized -> MusicAuthorizationStatus.Authorized
            initial -> MusicAuthorizationStatus.NotDetermined
            else -> MusicAuthorizationStatus.NotAuthorized
        }
    }

    private fun onPageEvent(name: String, data: JsonElement?) {
        when (name) {
            "playbackStateDidChange" -> {
                val stateName = BridgePayload.readString(data, "state")
                val state = PlaybackStateNames.tryParse(stateName)
                if (state == null) {
                    diagnose("MusicKit reported an unknown playback state '$stateName'")
                }

                diagnose("event playbackStateDidChange: ${stateName ?: "(none)"}")
                player.onStateChanged(state ?: MusicPlaybackState.None)
            }

            "nowPlayingItemDidChange" -> {
                val item = BridgePayload.readPayload<NowPlayingPayload>(data, "item")?.toModel()
                diagnose(
                    "event nowPlayingItemDidChange: ${item?.id ?: "(none)"} " +
                        "\"${item?.title ?: "(none)"}\"",
                )
                player.onNowPlayingChanged(item)
            }

            "queueItemsDidChange" ->
                diagnose("event queueItemsDidChange: ${BridgePayload.readInt(data, "count")} item(s)")

            "authorizationStatusDidChange" ->
                setAuthorization(BridgePayload.readBool(data, "isAuthorized"), initial = false)

            "mediaPlaybackError" -> {
                val message = BridgePayload.readString(data, "message")
                    ?: "Apple Music reported a playback error."
                diagnose("playback error: $message")
                player.onPlaybackError(message)
            }

            "error" -> diagnose("page error: ${BridgePayload.readString(data, "message")}")

            "musickitloaded" ->
                diagnose("MusicKit JS ${BridgePayload.readString(data, "version")} loaded")

            // Anything MusicKit adds later needs no typed handling.
            else -> Unit
        }
    }

    /**
     * The idle rotation. The seams — a queue being set, a play from a standstill, a skip — cover
     * every rotation a busy host needs; this covers the host that sat still for an hour and would
     * otherwise start its next set with a token that expires mid-song.
     */
    private fun startRotationTimer() {
        scope.launch {
            while (isActive) {
                delay(ROTATION_TICK)

                if (closed ||
                    !policy.shouldRotate(
                        RotationTrigger.IdleTimer,
                        player.state.value,
                        developerTokenExpiresAt,
                        Instant.now(),
                    )
                ) {
                    continue
                }

                try {
                    rotateNow()
                } catch (error: MusicKitException) {
                    diagnose("idle token rotation failed: ${error.message}")
                } catch (error: TimeoutException) {
                    diagnose("idle token rotation failed: ${error.message}")
                }
            }
        }
    }

    companion object {

        /** How often the idle timer considers rotating the developer token. */
        private val ROTATION_TICK = 60.seconds

        /** Apple's page limit for the library endpoints this uses. */
        private const val PAGE_SIZE = 100

        /** Enough pages for any real library; a cursor that never ends is a fault, not a big library. */
        private const val MAX_PAGES = 100

        /** How many lines are replayed to a new collector; a bounded buffer, not a transcript. */
        internal const val MAX_BUFFERED_LINES = 50

        private val UTC_TIME: DateTimeFormatter =
            DateTimeFormatter.ofPattern("HH:mm:ss").withZone(ZoneOffset.UTC)

        /**
         * Stands a host up: hidden WebView, bundled page, first developer token, configure.
         *
         * @param options where tokens come from, which activity to park the page in, and the timeouts.
         * @return a configured host, ready for catalog calls and — if the user is already signed in
         *   — playback.
         * @throws MusicKitException the WebView, the page, or the token could not be brought up.
         */
        suspend fun create(options: MusicKitOptions): MusicKitHost {
            // Built before anything that logs into it, so the lines startup produces are held by
            // the flow's own replay rather than dropped for want of a collector.
            val diagnostics = MutableSharedFlow<String>(
                replay = MAX_BUFFERED_LINES,
                extraBufferCapacity = 64,
                onBufferOverflow = BufferOverflow.DROP_OLDEST,
            )

            val webView = HostedWebView.create(options.activity) { diagnostics.tryEmit(it) }

            var host: MusicKitHost? = null
            try {
                val bridge = MusicKitBridge(webView, options.callTimeout, options.authorizeTimeout)
                val built = MusicKitHost(options, webView, bridge, diagnostics)
                host = built

                bridge.onLog = built::diagnose
                bridge.onPageEvent = built::onPageEvent
                webView.onMessage = bridge::onMessage
                webView.onRendererGone = bridge::failAll

                // A sign-in window that never appeared is as final for a waiting call as a dead
                // renderer, and is reported the same way: a retryable PAGE_GONE, now, rather than
                // an authorize() that hangs for three minutes over a popup nobody will ever see.
                webView.onAuthPopupFailed = bridge::failAll

                if (!webView.waitForPage(options.callTimeout)) {
                    throw MusicKitException(
                        "The Apple Music page did not finish loading. Check that the device can " +
                            "reach js-cdn.music.apple.com.",
                        "PAGE_NOT_READY",
                    )
                }

                val token = options.developerTokenProvider.getToken(forceRefresh = false)
                val configured = bridge.configure(token.token, options.appName, options.appBuild)

                built.developerTokenExpiresAt = token.expiresAt
                built.storefrontId = configured.storefrontId
                built.setAuthorization(configured.isAuthorized, initial = true)
                built.diagnose(
                    "MusicKit ${configured.version} configured; " +
                        "storefront ${configured.storefrontId ?: "(none)"}, " +
                        "authorized ${configured.isAuthorized}, " +
                        "token expires ${utc(token.expiresAt)}",
                )

                built.startRotationTimer()
                return built
            } catch (error: Throwable) {
                // Whatever went wrong is the answer; a failure while unwinding must not replace it.
                runCatching { host?.close() ?: webView.close() }
                throw error
            }
        }

        private fun utc(instant: Instant): String = "${UTC_TIME.format(instant)}Z"
    }
}

/** Unwraps a successful API answer, or turns an unsuccessful one into an exception. */
private fun requireBody(answer: ApiResult, what: String): JsonElement {
    val body = answer.body
    if (answer.isSuccess && body != null) {
        return body
    }

    val (message, code) = AppleJson.describeFailure(what, answer.status, answer.body, answer.error)
    throw MusicKitException(message, code)
}

File: musickit/src/main/assets/musickit/musickit-host.js
// The page half of the bridge. Kotlin sends {type:'invoke', id, method, args} over the channel
// androidx.webkit's addWebMessageListener injects as window.musicKitBridge; every call answers with
// exactly one {type:'result', id, ok, value|error} message. MusicKit's own events arrive unsolicited
// as {type:'event', name, data}. No Apple credential is ever posted back: the Music User Token is
// read here, used here, and never leaves the page.
(function () {
  'use strict';

  var app = { name: 'Druware.MusicKit', build: '0.1.0' };
  var music = null;

  // The only egress point from page to native. The WebView2 original used
  // window.chrome.webview.postMessage; the injected object androidx.webkit gives us is named
  // musicKitBridge and is reachable only from this origin, because addWebMessageListener was scoped
  // to it. Everything else about the protocol is unchanged.
  function post(payload) {
    if (window.musicKitBridge) {
      window.musicKitBridge.postMessage(JSON.stringify(payload));
    }
  }

  function log(message) { post({ type: 'log', message: String(message) }); }
  function emit(name, data) { post({ type: 'event', name: name, data: data || {} }); }
  function result(id, ok, value, error) {
    post({ type: 'result', id: id, ok: ok, value: ok ? (value === undefined ? null : value) : null, error: error || null });
  }

  function describe(error) {
    if (!error) { return 'unknown error'; }
    if (typeof error === 'string') { return error; }
    var parts = [];
    if (error.name) { parts.push(error.name); }
    if (error.message) { parts.push(error.message); }
    if (error.errorCode) { parts.push('errorCode=' + error.errorCode); }
    if (parts.length === 0) {
      try { return JSON.stringify(error); } catch (e) { return String(error); }
    }
    return parts.join(': ');
  }

  function stateName(value) {
    try {
      var names = window.MusicKit && window.MusicKit.PlaybackStates;
      return (names && names[value] !== undefined) ? String(names[value]) : String(value);
    } catch (e) { return String(value); }
  }

  // MusicKit hands API failures back in more than one shape depending on where they were raised,
  // so both the status and the body are dug for rather than assumed.
  function statusOf(error) {
    if (!error) { return null; }
    if (typeof error.status === 'number') { return error.status; }
    if (typeof error.httpStatusCode === 'number') { return error.httpStatusCode; }
    if (error.response && typeof error.response.status === 'number') { return error.response.status; }
    // MusicKit raises its own API failures as { name: 'NOT_FOUND', message: '404' }, where the
    // status is only ever the message. Without this a 404 — which is an answer — reaches the host
    // as an unexplained failure instead of "no such thing".
    if (/^[1-5][0-9][0-9]$/.test(String(error.message))) { return Number(error.message); }
    return null;
  }

  function bodyOf(error) {
    if (!error) { return null; }
    if (error.data && error.data.errors) { return error.data; }
    if (error.body && error.body.errors) { return error.body; }
    if (error.errors) { return { errors: error.errors }; }
    return null;
  }

  function playParamsOf(item) {
    if (!item) { return null; }
    return item.playParams || (item.attributes && item.attributes.playParams) || null;
  }

  function summarise(item) {
    if (!item) { return null; }
    var a = item.attributes || item;
    var pp = playParamsOf(item) || {};
    var isSong = pp.kind === undefined || pp.kind === 'song';
    return {
      id: item.id || pp.id || null,
      title: a.name || a.title || item.title || null,
      artistName: a.artistName || item.artistName || null,
      catalogId: pp.catalogId || (isSong && !pp.isLibrary ? (pp.id || null) : null),
      libraryId: (isSong && pp.isLibrary) ? (pp.id || null) : null
    };
  }

  function requireInstance() {
    if (!music) { throw new Error('MusicKit is not configured yet.'); }
    return music;
  }

  var audioUnlocked = false;

  // Chromium will not let MusicKit's <audio> element stream in a document whose audio has never
  // been started, and a WebView nobody can see never gets the gesture that would start it. The
  // failure is silent and total: MusicKit accepts the queue, reports the now-playing item, goes
  // playing -> waiting -> loading, and stays in loading. play() does not settle for a minute and a
  // half, no mediaPlaybackError is ever raised, and every later call to that instance hangs behind
  // it. Rendering one near-silent oscillator through an AudioContext first is what avoids all of
  // that — proved by pairs of live runs that differ in nothing else.
  async function unlockAudio() {
    if (audioUnlocked) { return; }
    audioUnlocked = true;
    try {
      var ctx = new AudioContext();
      if (ctx.state === 'suspended') { await ctx.resume(); }

      var oscillator = ctx.createOscillator();
      var gain = ctx.createGain();
      gain.gain.value = 0.0001;
      oscillator.connect(gain);
      gain.connect(ctx.destination);
      oscillator.start();

      // The clock only advances while audio is really being rendered, so a second of it is the
      // proof as well as the unlocking. A hidden page's timers are throttled to about a tick a
      // second, which is why this waits on one timer rather than polling.
      await new Promise(function (resolve) { setTimeout(resolve, 1000); });
      log('audio unlocked: AudioContext ' + ctx.state + ', clock at ' + ctx.currentTime.toFixed(3) + ' s');

      try { oscillator.stop(); } catch (e) { /* already stopped */ }
      await ctx.close();
    } catch (error) {
      audioUnlocked = false;
      log('audio could not be unlocked, so playback may never leave loading: ' + describe(error));
    }
  }

  // Attached once per instance: a reconfigure hands back the same singleton, and listeners added
  // twice would double every event.
  function attach(instance) {
    if (!instance || instance.__druwareAttached) { return instance; }
    instance.__druwareAttached = true;

    instance.addEventListener('playbackStateDidChange', function (e) {
      emit('playbackStateDidChange', { state: stateName(e && e.state !== undefined ? e.state : instance.playbackState) });
    });
    instance.addEventListener('nowPlayingItemDidChange', function (e) {
      emit('nowPlayingItemDidChange', { item: summarise((e && e.item) || instance.nowPlayingItem) });
    });
    instance.addEventListener('authorizationStatusDidChange', function () {
      emit('authorizationStatusDidChange', { isAuthorized: !!instance.isAuthorized });
    });
    instance.addEventListener('queueItemsDidChange', function (items) {
      emit('queueItemsDidChange', { count: (items && items.length) ? items.length : 0 });
    });
    instance.addEventListener('mediaPlaybackError', function (e) {
      emit('mediaPlaybackError', { message: describe(e) });
    });

    return instance;
  }

  var announced = false;
  function announce() {
    if (announced) { return; }
    announced = true;
    emit('musickitloaded', { version: (window.MusicKit && window.MusicKit.version) || null });
  }

  if (window.MusicKit) { announce(); }
  document.addEventListener('musickitloaded', announce);

  window.onerror = function (message, source, line, column, error) {
    emit('error', { message: String(message) + ' @' + line + ':' + column, detail: error ? describe(error) : null });
  };
  window.addEventListener('unhandledrejection', function (e) {
    emit('error', { message: describe(e.reason) });
  });

  // One Apple Music API call. Never throws for an API failure: the status and the body go back so
  // the host can tell a 404 (which is an answer) from a 403 (which is not).
  async function apiCall(path, params) {
    var instance = requireInstance();
    try {
      var response = await instance.api.music(path, params || undefined);
      return {
        status: (response && response.response && response.response.status) || 200,
        body: (response && response.data) || null,
        error: null
      };
    } catch (e) {
      return { status: statusOf(e), body: bodyOf(e), error: describe(e) };
    }
  }

  var handlers = {
    status: function () {
      return {
        musicKitLoaded: !!window.MusicKit,
        version: (window.MusicKit && window.MusicKit.version) || null,
        configured: !!music,
        isAuthorized: !!(music && music.isAuthorized),
        storefrontId: music ? (music.storefrontId || null) : null
      };
    },

    configure: async function (args) {
      if (!window.MusicKit) { throw new Error("MusicKit JS has not loaded from Apple's CDN."); }
      if (args.appName) { app = { name: args.appName, build: args.appBuild || '0.1.0' }; }
      await window.MusicKit.configure({ developerToken: args.token, app: app });
      music = attach(window.MusicKit.getInstance());
      await unlockAudio();
      return {
        version: window.MusicKit.version || null,
        storefrontId: music.storefrontId || null,
        isAuthorized: !!music.isAuthorized
      };
    },

    // Rotation. MusicKit exposes developerToken as a getter only, so a second configure is the only
    // way to hand it a new one. It returns the same singleton, keeps the sign-in, and stops
    // playback — the host re-applies the queue afterwards.
    reconfigure: async function (args) {
      if (!window.MusicKit) { throw new Error("MusicKit JS has not loaded from Apple's CDN."); }
      await window.MusicKit.configure({ developerToken: args.token, app: app });
      music = attach(window.MusicKit.getInstance());
      return {
        version: window.MusicKit.version || null,
        storefrontId: music.storefrontId || null,
        isAuthorized: !!music.isAuthorized,
        developerTokenMatches: music.developerToken === args.token
      };
    },

    // authorize() resolves to the Music User Token. It is deliberately not returned: only the
    // boolean crosses the bridge, so no Apple credential ever reaches Kotlin or a log.
    authorize: async function () {
      var instance = requireInstance();
      await instance.authorize();
      return { isAuthorized: !!instance.isAuthorized };
    },

    unauthorize: async function () {
      var instance = requireInstance();
      await instance.unauthorize();
      return { isAuthorized: !!instance.isAuthorized };
    },

    apiGet: function (args) {
      return apiCall(args.path, args.params);
    },

    // Creating a library playlist needs a POST, and music.api.music offers none: its third argument
    // takes a fetchOptions object, but MusicKit 3.2526 ignores the method in it. The probe in the
    // integration test asks a catalog song for a POST and gets 200 with the song's data back — a
    // GET. So the request is made with a plain fetch here in the page, where the Music User Token
    // already is and so where it stays. Re-run that probe before trusting a newer MusicKit.
    apiPost: async function (args) {
      var instance = requireInstance();
      var response = await fetch('https://api.music.apple.com' + args.path, {
        method: 'POST',
        headers: {
          'Authorization': 'Bearer ' + instance.developerToken,
          'Music-User-Token': instance.musicUserToken,
          'Content-Type': 'application/json'
        },
        body: args.body
      });

      var text = await response.text();
      var parsed = null;
      if (text) { try { parsed = JSON.parse(text); } catch (e) { parsed = null; } }

      return {
        status: response.status,
        body: parsed,
        error: response.ok ? null : ('HTTP ' + response.status + ' ' + response.statusText),
        via: 'fetch'
      };
    },

    // 'items' names each identifier's type and carries its play parameters, which is what lets
    // MusicKit build the queue from the descriptor alone — no Apple request, and so nothing to
    // rate-limit or fail to resolve. An item without playParams sends MusicKit to
    // /v1/me/library/undefineds/…, so the parameters are not optional.
    setQueue: async function (args) {
      var instance = requireInstance();
      var started = Date.now();
      log('setQueue entry: ' + args.ids.length + ' id(s), state ' + stateName(instance.playbackState));
      var descriptor = {
        items: args.ids.map(function (id) {
          var isLibrary = String(id).indexOf('i.') === 0 || String(id).indexOf('l.') === 0;
          var playParams = { id: id, kind: 'song' };
          if (isLibrary) { playParams.isLibrary = true; }
          return {
            id: id,
            type: isLibrary ? 'library-songs' : 'songs',
            attributes: { playParams: playParams }
          };
        })
      };

      await instance.setQueue(descriptor);
      var queue = instance.queue;
      var items = queue && queue.items ? queue.items : [];
      log('setQueue exit: ' + items.length + ' item(s), state ' + stateName(instance.playbackState) +
        ', ' + (Date.now() - started) + ' ms');
      return { count: items.length };
    },

    play: async function () {
      var instance = requireInstance();
      var started = Date.now();
      log('play entry: state ' + stateName(instance.playbackState) +
        (instance.nowPlayingItem ? '' : ', nowPlayingItem is null'));
      await instance.play();
      log('play exit: state ' + stateName(instance.playbackState) + ', ' + (Date.now() - started) + ' ms');
      return { playbackState: stateName(instance.playbackState) };
    },

    pause: async function () {
      var instance = requireInstance();
      await instance.pause();
      return { playbackState: stateName(instance.playbackState) };
    },

    stop: async function () {
      var instance = requireInstance();
      await instance.stop();
      return { playbackState: stateName(instance.playbackState) };
    },

    skipToNext: async function () {
      var instance = requireInstance();
      await instance.skipToNextItem();
      return { playbackState: stateName(instance.playbackState), item: summarise(instance.nowPlayingItem) };
    },

    nowPlaying: function () {
      var instance = requireInstance();
      return { playbackState: stateName(instance.playbackState), item: summarise(instance.nowPlayingItem) };
    },

    // Reports whether this WebView build can start Widevine EME at all. Android WebView, unlike
    // Chrome, only gets a CDM when the host app grants PROTECTED_MEDIA_ID in onPermissionRequest,
    // and some OEM WebView forks carry no Widevine CDM whatsoever — so this is a capability probe,
    // not an assertion. Never throws: a rejection is an answer.
    probeWidevine: async function () {
      if (!navigator.requestMediaKeySystemAccess) {
        return { supported: false, error: 'requestMediaKeySystemAccess is not available.' };
      }

      try {
        var access = await navigator.requestMediaKeySystemAccess('com.widevine.alpha', [{
          initDataTypes: ['cenc'],
          audioCapabilities: [{ contentType: 'audio/mp4; codecs="mp4a.40.2"' }]
        }]);
        return { supported: true, keySystem: (access && access.keySystem) || null, error: null };
      } catch (error) {
        return { supported: false, keySystem: null, error: describe(error) };
      }
    }
  };

  window.musicKitHost = {
    invoke: function (id, method, args) {
      try {
        var handler = handlers[method];
        if (!handler) { result(id, false, null, "unknown bridge method '" + method + "'"); return; }
        Promise.resolve(handler(args || {})).then(
          function (value) { result(id, true, value, null); },
          function (error) { result(id, false, null, describe(error)); });
      } catch (error) {
        result(id, false, null, describe(error));
      }
    }
  };

  // The native half's only way in. WebView2 could hand the page a script to run; androidx.webkit's
  // channel is symmetric instead, so an invocation arrives here as a message rather than as an
  // evaluated expression — which is why nothing native ever needs to build JS source at run time.
  if (window.musicKitBridge) {
    window.musicKitBridge.onmessage = function (event) {
      var call = null;
      try { call = JSON.parse((event && event.data) || ''); } catch (e) { return; }
      if (!call || call.type !== 'invoke') { return; }
      window.musicKitHost.invoke(call.id, call.method, call.args);
    };
  }

  log('musickit-host.js ready; MusicKit ' + (window.MusicKit ? 'already present' : 'not yet loaded') +
    ', channel ' + (window.musicKitBridge ? 'attached' : 'MISSING'));
})();
